accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -1137,16 +1143,21 @@ arguments) marked as nullable or non-nullable as written.
 If a class `C` in an opted-in library implements the same generic class `I` more
 than once as `I0, .., In`, and at least one of the `Ii` is not syntactically
 equal to the others, then it is an error if `NNBD_TOP_MERGE(S0, ..., Sn)` is not
-defined where `Si` is **NORM(`Ii`)**.  Otherwise, for the purposes of runtime
-subtyping checks, `C` is considered to implement the canonical interface given
-by `NNBD_TOP_MERGE(S0, ..., Sn)`.
+defined where `Si` is **NORM(`Ii`)**.
+
+If `C` implements `I` once, or `C` implements `I` multiple times without error,
```

Comment:
This means that for
```Dart
  // opt-out
  class A implement Foo<FutureOr<Object>> {}
  // opt-in
  class B implements A {} 
```
`A` implements `Foo<FutureOr<Object>>` whereas `B` implements `Foo<Future<Object>>`. Is this intentional?

Reply:
I'm not keen on this change.  My goal here was to only invent a new type if required.  Why this change?

Reply:
For the example, normalization would reduce `Foo<FutureOr<Object*>*>` to `Foo<Object*>`, which eliminates `FutureOr`, and because `Object**` normalizes to `Object*`. So `A` implements `Foo<Object*>`.

And `B` implements `Foo<Object*>` because that's the `LEGACY_ERASURE` of `Foo<Object*>` (actually I think that erasure is `Foo<Object*>*`, but I don't think we want a rule that makes implemented types legacy at the top level, so we may need a variant of `LEGACY_ERASURE` that doesn't add a `*` at the top level, or maybe `LEGACY_ERASURE` just shouldn't add `*` at the top level at all).

But, @leafpetersen, I understand that you view this rule as particularly aggressive in its introduction of legacy typing. I used that wording in order to obtain a property that @johnniwinther mentioned a while back: A legacy class is purely legacy.

There may be different trade-offs here ("keep null-safe types whenever possible", in particular when `I` is only implemented once, versus "make sure that all properties of a legacy class are expressed in terms of legacy types").

@johnniwinther, WDYT?

Reply:
For me difference is in whether we should perform normalization or not. If there is only one type we don't need to merge and therefore just use the type itself. In this case that'll make `B` implement `Foo<FutureOr<Object*>*>`. If there are more types we need to perform `NNBD_TOP_MERGE` and in this case we perform `NORM` on the input:
```Dart
// opt-out
class A implements Foo<FutureOr<Object>> {}
// opt-in
class B implements Foo<FutureOr<dynamic>> {}
class C implements A, B {}
```
Here `C` implements 
```
NNBD_TOP_MERGE(NORM(Foo<FutureOr<Object*>*>),NORM(Foo<FutureOr<dynamic>>))
  = NNBD_TOP_MERGE(Foo<Object*>,Foo<dynamic>)
  = Foo<Object?>
```

Note that I don't consider the outmost declaration in extends, implements and with clauses to have nullability so `Foo` is just `Foo` regardless of opt-in/out status.

Reply:
@eernstg I don't understand your answer.  The issue that @johnniwinther was pointing out and that I was responding to has nothing to do with legacy, as far as I can tell.  My original spec only applied NNBD_TOP_MERGE when there were multiple super-interfaces with the same generic base class.  You add the the line `If `C` implements `I` once,`.  This line is the one that I question.  It implies we normalize all super-interfaces, which is a much higher impact change.

Reply:
The "when `C` implements `I` once" change should only be applied when `C` is legacy, it was a mistake to have it here. Sorry about that.

Reply:
However, I would expect an implementation to be allowed to normalize _any_ term derived from `<type>` as eagerly as it wishes. Some of these terms may have syntax like `FutureOr<X>` where `X` is a type variable, and this may obviously call for normalization at run time, but with `FutureOr<Object?>` I would expect it to be OK for a compiler to normalize it at compile-time.

@johnniwinther, do you prefer to avoid compile-time type normalization except in the case where it is used to resolve a conflict among implemented superinterfaces or inherited member signatures? I guess the point would be that normalization which is eager to some unknown/unspecified degree would create problems in the management of the values of constant expressions? (Some constant expressions suddenly evaluate to the same object, even though they look different).

I think that issue might not arise with the normalization that is applied during the two cases of conflict resolution that are in focus here, because canonicalization would only ever identify instances of the same class, and if we have two instances of the given class `C` then it makes no difference that they implement a normalized version of some superinterface, because that will be true for both of them.

Reply:
I prefer to only normalize when we need to synthesize a result (e.g. in case of implementing to instantiations of a generic interface).

Reply:
I think there are no further issues here: Normalizing when a synthesized result is needed is necessary, and if we wish to introduce normalization elsewhere then it would be a non-breaking change. So I'll resolve this thread.